tags/0.9.8-rc3/daemon/auth/btgpasswd.man
.TH BTGPASSWD 1 "16 May 2006"
.SH NAME
btgpasswd \- application for managing btg daemon password file.
.SH SYNOPSIS
.B "btgpasswd" 
[-a] [-m] [-r] [-l]
[-t 
.I "directory"
]
[-w 
.I "directory"
]
[-s 
.I "directory"
]
[-d 
.I "directory"
]
[-p 
.I "password"
]
[-u 
.I "username"
]
[--control 
.I "1"
| --control 
.I "0"
]
[-c | --create]
[-n]
[-f 
.I "FILE"
|--config 
.I "FILE"
]
[-v] 
[-h]

.SH DESCRIPTION
.B "btgpasswd" 
manages the password file used by the btg daemon. It supports adding, modifying, removing and listing entries.

.B "btgpasswd" 
reads the btg daemon config file, to find out which password file should be used. 
.B It will not work without it.

.SH OPTIONS
.TP
.BI "-a"
\- add a new entry.
.TP
.BI "-m"
\- modify an entry.
.TP
.BI "-r"
\- remove an entry.
.TP
.BI "-l"
\- List entries.

.TP
.BI "-t directory"
\- Use temp 
.I directory.

.TP
.BI "-w directory"
\- Use work 
.I directory.

.TP
.BI "-s directory"
\- Use seed 
.I directory.

.TP
.BI "-d directory"
\- Use destination 
.I "directory".

.TP
.BI "-p password"
\- Use the argument as the password.

.TP
.BI "-u username"
\- Use the argument as the username.

.TP
.BI "--control 1|--control 0"
\- Enable/disable the control flag. It allows an user to: shutdown btgdaemon, set global limits.

.TP
.BI "-c | --create"
\- Create a new file. 
.BI "This is to be used in connection with the -a switch, for creating new users."

.TP
.BI "-f filename | --config filename"
\- Specify the daemon configuration file to use.

.TP
.BI "-v"
\- Print the version.
.TP
.BI "-h|--help"
\- Print a short help.

.SH EXAMPLES
.TP 
Creating a new file (Done by adding a new user).
btgpasswd --create -a -u username -p password -d /path/to/dest -w /path/to/work -s /path/to/seed -t /path/to/temp
.TP
Adding a new user (passwd file exists).
btgpasswd -a -u username2 -p password2 -d /path/to/dest -w /path/to/work -s /path/to/seed -t /path/to/temp

.SH FILES
.TP
~/.btg/daemon.ini - default daemon configuration file is used, unless --config is used to provide another daemon configuration file.

.SH AUTHOR
Written by Michael Wojciechowski.

.SH "SEE ALSO"
btg(5), btg_config(5)

.SH BUGS
No known bugs.


trunk/ChangeLog
BTG version 0.9.8
-----------------
The following bugs were fixed:

014142 Compilation broke on curl link with --as-needed LDFLAG.
013961 race-condition in btgui (in peer-list)
014067 BTG SVN fails to compile with SVN Libtorrent.
014061 XMLRPC_Client fails to parse some XMLs.
013987 btgcli can't read commands from file.
013985 Installed xmlrpc-epi should be tested.
013984 Remove bundled expat.
013983 Installed plotmm should be tested.
013975 Incorrect torrent statuses.
013971 Display build and revision numbers on the wwwUI pages.
013964 btgcli: Do not print LF when displaying downloaded pieces.
013932 btg-daemon-0.9.8 not installable on debian.
013931 Bad URL displayed after DEB install.
013859 Remove ncurses from btgcli.
013852 Use torrent-ip to create lt sessions.
013833 Convert projectDefaults to a non-singleton.
013832 XMLRPC response parse error.
013814 Configure script does not detect lt upnp in lt 0.14.
013774 Add amd64 architecture to the packages.
013720 Daemon hangup on incorrect torrent file.
013708 Seeder/Leecher counts incorrect.
013701 Keys for "Quit" and "Back" should be separate.
013700 Reload filter list w/o restarting daemon.
013678 Update openwrt makefiles.
013645 Smooth URL loading indicator.
013644 Ports opened by UPnP when it is disabled.
013643 Daemon starts without rewriting unexistent pid.
013642 btgui incorrect behavior during init. when comm. error.
013618 btgncli, show sessions when invoked without arguments.
013557 CyberLink UPnP, automatic internal IP detection.
013555 UPnP, added separate config-file parameter for internal-ip.
013538 Issues with native FreeBSD dialog.
013501 Implement better torrent loading.
013490 Provide source-packages in deb-src.
013460 Showing file pieces for torrents missing from wwwui.
013459 File selection missing from wwwui.
013454 Writing PID-file.
013450 Logging optimization - static conditions.
013436 btgui, resizable headers.
013435 Open all necessary files before switching EUID (daemon).
013434 Log messsages loss when logging not initialized.
013433 Configuration file search paths.
013430 Saving opened urls in history.
013412 Additional command: supported options.
013411 BTGWWW, load a torrent from an URL.
013399 Merge Miklos Vajna's GCC 4.3 patch.
013381 URL loading in daemon and non-www clients.
013307 btgui, no error message on incorrect login (also btgcli/ncli).
013198 Unable to load huge torrent files.

BTG version 0.9.7
-----------------
The following bugs were fixed:

013359 Compilation error: numinput.h.
013315 btgui, torrent remains after moving to another session.
013306 btgui, executing operations on multiple torrents.
013272 Add support for libtorrent 0.12.1.
013267 btgcli, incorrect seed limit.
013266 btgui, global limit setting.
013265 btgui, limit settings dialog doesn't correctly display limits.
013248 btgncli, add dialog used for entering numbers.
013216 Session saving, add saving of global limits.
013147 Handle octets set to 255 in Address/addressPort.
013146 XML Pasing error (wwwui).
013142 Restore file priorities after restart.
013136 Change of peer id / useragent (feature req).
013007 Show session settings.
012997 Session saving, fix saving of dht/encryption settings.
012975 Externalization, simple, exorcism.
012970 Add moving torrents to another sessions.
012952 Add switch to disable generating btg-config.
012839 DHT + encryption support in wwwbtg.
012814 Daemon segfaults when starting client with DHT (0.9.6).
012784 Remove singleton object used for logging.
012771 Add openwrt packages to SVN.
012762 Embed svn version into release information.
012761 Callback manager: execute calls from a thread.
012685 btgdaemon/btgpassswd segfaulting upon start.
012676 Fix boost iostreams detection.
012616 Remove btgq and btgd.
012598 Write man page for btgvs.
012557 Make fails on FreeBSD 6.2.
012545 Update SDL viewer to AGAR 1.3.1.
012544 BTGWWW, add missing files in bcore/command.
012535 btgncli: show peers/seeds in the list of torrents.
012330 Add another version of contextStatusCommand.
012329 Add "stop autoupdating" button in wwwbtg.
012308 Clean up daemon/limitval.cpp.
012301 Update doxygen.
012241 Use pkglib_LTLIBRARIES instead of lib_LTLIBRARIES.
012226 btgui, files showing as 100%.
012208 Add logging using syslog to wwwui.
012205 torrent files which contain same files but different tracker.
012190 Fix "warning: 'add_torrent' is deprecated".
012125 Update SDL viewer, after refactoring of last files.
012124 Add command line interface(s) unittest.
012120 Update to libtorrent 0.13 aka SVN.
012118 Add session naming.
012108 btgui: buttons OK and Cancel are equivalent when logging-in.
012106 Can't set files to download in btgui.
012055 File checking fails on torrents that contain directories.
012004 Show tracker for each context.
011784 More users can't download same torrent at same time.
011781 FBSD, btgncli does not show the status bar.
011557 Peer_id sent to tracker (feature req).

BTG version 0.9.6
-----------------
The following bugs were fixed:

011971 Inaccuracy in command (Loading All Torrent Files from a Dir).
011902 Memory leak in external/ini.
011901 FBSD, memory leak when using boost/zlib.
011782 FBSD, btgncli uses ~60% of cpu.
011759 Update howto and configure script.
011719 WWWUI does not respond (hardcoded port).
011678 Add more XML-RPC unittests.
011674 Update XML-RPC in external.
011644 WebBTG not working with Internet Explorer.
011575 / 010064 daemon should check "seed" dir for download files.
011440 btgui, use a thread for updates.
011237 wwwui, unable to handle "+" in passwords.
011192 Document release procedure.
011187 DHT, add command line argument to enable/disable.
011186 Implement support for encryption.
011185 Implement support for partial downloads.
011184 Remove use of old libtorrent versions.
011183 Session saving, use xmlrpc.
011122 Update site/howto: Gentoo ebuild	documentation.
011097 btgpasswd, check for unique directories (request).

branches/stable-branch/daemon/auth/btgpasswd.cpp
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id$
 */

#include <bcore/project.h>
#include <bcore/logmacro.h>

#include "arg.h"
#include <daemon/dconfig.h>

#include <bcore/os/fileop.h>

#include "passwd_auth.h"

#include <bcore/auth/hash.h>

#include <bcore/os/stdin.h>

#include <bcore/verbose.h>

#include <algorithm>

using namespace btg::core;
using namespace btg::core::logger;
using namespace btg::daemon::auth;

bool createDirectory(LogWrapperType _logwrapper, bool _verboseFlag, std::string const& _label, std::string const& _path);

bool uniqueDirectories(LogWrapperType _logwrapper,
                       std::string const& _temp_dir,
                       std::string const& _work_dir,
                       std::string const& _seed_dir,
                       std::string const& _dest_dir);

int main(int argc, char* argv[])
{
   LogWrapperType logwrapper(new btg::core::logger::logWrapper);

   commandLineArgumentHandler* cla = new commandLineArgumentHandler(projectDefaults::sDAEMON_CONFIG());
   cla->setup();
   if (!cla->parse(argc, argv))
      {
         delete cla;
         cla = 0;

         return BTG_ERROR_EXIT;
      }

   bool verboseFlag = cla->beVerbose();

   std::string configFile = projectDefaults::sDAEMON_CONFIG();
   if (cla->configFileSet())
      {
         configFile = cla->configFile();
      }

   VERBOSE_LOG(logwrapper, verboseFlag, "config: '" << configFile << "'.");

   std::string errorString;
   if (!btg::core::os::fileOperation::check(configFile, errorString, false))
      {
         BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to open config file '" << configFile << "': " << errorString);

         delete cla;
         cla = 0;

         return BTG_ERROR_EXIT;
      }

   btg::daemon::daemonConfiguration* config  = new btg::daemon::daemonConfiguration(logwrapper,
                                                                                    configFile);

   if (!config->read())
      {
         BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Could not open configuration file '" << configFile << "'");

         delete config;
         config = 0;
         delete cla;
         cla = 0;

         return BTG_ERROR_EXIT;
      }

#if BTG_AUTH_DEBUG
   BTG_NOTICE(logwrapper, "Config read from '" << configFile << "'.");
#endif // BTG_AUTH_DEBUG

   // Open the passwd file, filename from the config file.

   std::string auth_file = config->getAuthFile();

   bool newfile = cla->createNewFile();

   if (newfile)
      {
         if (cla->getOperation() != commandLineArgumentHandler::OP_ADD)
            {
               // Only allow creating new file when an user is being added.
               newfile = false;
            }
      }

   // resolve relative paths etc.
   if (!btg::core::os::fileOperation::resolvePath(auth_file))
      {
         BTG_NOTICE(logwrapper, "Unable to resolve path '" << auth_file << "'");
      }

   if (!newfile)
      {
         if (!btg::core::os::fileOperation::check(auth_file, errorString, false))
            {
               BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Could not open passwd file '" << auth_file << "': " << errorString);

               delete config;
               config = 0;
               delete cla;
               cla = 0;

               return BTG_ERROR_EXIT;
            }
      }

   passwordAuth* auth = new passwordAuth(logwrapper, auth_file, newfile);

   if (!auth->initialized())
      {
         BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to load passwd file '" << auth_file << "'.");

         delete auth;
         auth = 0;
         delete config;
         config = 0;
         delete cla;
         cla = 0;

         return BTG_ERROR_EXIT;
      }

   commandLineArgumentHandler::OPERATION operation = cla->getOperation();

   switch (operation)
      {
      case commandLineArgumentHandler::OP_ADD:
         {
            VERBOSE_LOG(logwrapper, verboseFlag, "Adding user.");

            std::string user;
            std::string password;
            std::string password_confirm;
            std::string temp_dir;
            std::string work_dir;
            std::string seed_dir;
            std::string dest_dir;
            bool        control_flag = false;
            std::string callback;

            if (!cla->getUsername(user))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing user name.");
                  break;
               }

            if (!cla->getTempDir(temp_dir))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing temporary directory.");
                  break;
               }

            if (!cla->getWorkDir(work_dir))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing working directory.");
                  break;
               }

            if (!cla->getSeedDir(seed_dir))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing seed directory.");
                  break;
               }

            if (!cla->getDestDir(dest_dir))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing destination directory.");
                  break;
               }

            // Make sure that the directories are unique.
            if (!uniqueDirectories(logwrapper,
                                   temp_dir, work_dir,
                                   seed_dir, dest_dir))
               {
                  BTG_FATAL_ERROR(logwrapper,
                                  "btgpasswd", "Provided directories are not unique. " << 
                                  "Make sure temporary/working/seed/destination directory is unique.");
                  break;
               }

            if (!cla->getControlFlag(control_flag))
               {
                  control_flag = false;
               }

            if (!cla->getCallback(callback))
               {
                  callback = Auth::callbackDisabled;
               }

            std::cout << "Enter password:" << std::endl;

            // Ask for a password.
            if (!btg::core::os::stdIn::getPassword(password))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Reading password failed.");
                  break;
               }

            std::cout << "Confirm password:" << std::endl;

            // Ask for a password.
            if (!btg::core::os::stdIn::getPassword(password_confirm))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Reading password failed.");
                  break;
               }

            // Make sure passwords are equal
            if (password != password_confirm)
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Password missmatch.");
                  break;
               }


            Hash h;
            h.generate(password);
            std::string hashed;
            h.get(hashed);

            VERBOSE_LOG(logwrapper,
                        verboseFlag, "Adding user '" << user << "' with:" << "\n" <<
                        "hash '" << hashed << "'" << "\n" <<
                        "temp directory '" << temp_dir << "'" << "\n" <<
                        "work directory '" << work_dir << "'" << "\n" <<
                        "seed directory '" << seed_dir << "'" << "\n" <<
                        "destination directory '" << dest_dir << "'" << "\n" << 
                        "control flag '" << control_flag << "'" << "\n" <<
                        "callback '" << callback << "'" << "\n"
                        );

            // Create the directories used.

            if ((!createDirectory(logwrapper, verboseFlag, "temp", temp_dir)) ||
                (!createDirectory(logwrapper, verboseFlag, "work", work_dir)) ||
                (!createDirectory(logwrapper, verboseFlag, "seed", seed_dir)) ||
                (!createDirectory(logwrapper, verboseFlag, "dest", dest_dir))
                )
               {
                  // Abort, as some directories failed.
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Failed to create user directorys.");
                  break;
               }

            if (!auth->addUser(user, hashed, temp_dir, work_dir, seed_dir, dest_dir, control_flag, callback))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to add user.");
                  break;
               }

            if (!auth->write())
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to save password file.");
                  break;
               }

            std::cout << "User '" << user << "' added." << std::endl;
            break;
         }
      case commandLineArgumentHandler::OP_MOD:
         {
            VERBOSE_LOG(logwrapper, verboseFlag, "Modifying user.");

            std::string user;
            std::string password;
            std::string password_confirm;
            std::string temp_dir;
            std::string work_dir;
            std::string seed_dir;
            std::string dest_dir;

            bool control_flag = false;

            std::string callback;

            if (!cla->getUsername(user))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing user name.");
                  break;
               }

            if (!auth->getControlFlag(user, control_flag))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing control flag.");
                  break;
               }

            bool any_argument = false;

            if (cla->readPassword())
               {
                  any_argument = true;
               }

            if (cla->getTempDir(temp_dir))
               {
                  any_argument = true;
               }

            if (cla->getWorkDir(work_dir))
               {
                  any_argument = true;
               }

            if (cla->getSeedDir(seed_dir))
               {
                  any_argument = true;
               }
            if (cla->getDestDir(dest_dir))
               {
                  any_argument = true;
               }

            if (cla->getControlFlag(control_flag))
               {
                  any_argument = true;
               }

            if (cla->getCallback(callback))
               {
                  any_argument = true;
               }

            if (!any_argument)
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Nothing changed.");
                  break;
               }

            BTG_NOTICE(logwrapper,
                       "Mod user '" << user << "', using directories:" << "\n" <<
                       "temp directory '" << temp_dir << "'" << "\n" <<
                       "work directory '" << work_dir << "'" << "\n" <<
                       "seed directory '" << seed_dir << "'" << "\n" <<
                       "destination directory '" << dest_dir << "'" << "\n"
                       );
            std::string hashedPassword;
            if (cla->readPassword())
               {
                  std::cout << "Enter password:" << std::endl;

                  // Ask for a password.
                  if (!btg::core::os::stdIn::getPassword(password))
                     {
                        BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Reading password failed.");
                        break;
                     }

                  std::cout << "Confirm password:" << std::endl;

                  // Ask for a password.
                  if (!btg::core::os::stdIn::getPassword(password_confirm))
                     {
                        BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Reading password failed.");
                        break;
                     }

                  // Make sure passwords are equal
                  if (password != password_confirm)
                     {
                        BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Password missmatch.");
                        break;
                     }

                  Hash h;
                  h.generate(password);
                  h.get(hashedPassword);
               }

            if (verboseFlag)
               {
                  std::string output;

                  output += "Modyfying user '" + user + "', setting:" + "\n";

                  if (password.size() > 0)
                     {
                        output += "hash '" + hashedPassword + "'" + "\n";
                     }

                  if (temp_dir.size() > 0)
                     {
                        output += "temp directory '" + temp_dir + "'" + "\n";
                     }

                  if (work_dir.size() > 0)
                     {
                        output += "work directory '" + work_dir + "'" + "\n";
                     }

                  if (seed_dir.size() > 0)
                     {
                        output += "seed directory '" + seed_dir + "'" + "\n";
                     }

                  if (dest_dir.size() > 0)
                     {
                        output += "destination directory '" + dest_dir + "'" + "\n";
                     }

                  if (control_flag)
                     {
                        output += "control flag '1'\n";
                     }
                  else
                     {
                        output += "control flag '0'\n";
                     }

                  if (callback.size() > 0)
                     {
                        output += "callback '" + callback + "'\n";
                     }

                  VERBOSE_LOG(logwrapper, verboseFlag, output);
               } // verbose

            if (!auth->modUser(user, hashedPassword, temp_dir, work_dir, seed_dir, dest_dir, control_flag, callback))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to modify user.");
                  break;
               }

            if (!auth->write())
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to save password file.");
                  break;
               }

            std::cout << "User '" << user << "' modified." << std::endl;
            break;
         }
      case commandLineArgumentHandler::OP_DEL:
         {
            VERBOSE_LOG(logwrapper, verboseFlag, "Deleting user.");

            std::string user;

            if (!cla->getUsername(user))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Missing user name.");
                  break;
               }

            if (!auth->delUser(user))
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to delete user.");
                  break;
               }

            if (!auth->write())
               {
                  BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unable to save password file.");
                  break;
               }

            std::cout << "Deleted user '" << user << "'." << std::endl;
            break;
         }
      case commandLineArgumentHandler::OP_LST:
         {
            std::cout << "Listing users." << std::endl;
            std::cout << auth->toString() << std::endl;
            break;
         }
      default:
         {
            BTG_FATAL_ERROR(logwrapper, "btgpasswd", "Unknown operation.");
         }
      }

   delete auth;
   auth = 0;

   delete config;
   config = 0;

   delete cla;
   cla = 0;

   return 0;
}

bool createDirectory(LogWrapperType _logwrapper,
                     bool _verboseFlag, std::string const& _label, std::string const& _path)
{
   bool result = btg::core::os::fileOperation::createDirectory(_path);

   switch(result)
      {
      case true:
         VERBOSE_LOG(_logwrapper, _verboseFlag, "Created " << _label << " directory.");
         break;
      case false:
         VERBOSE_LOG(_logwrapper, _verboseFlag, "Unable to create " << _label << " directory, path .");
         break;
      }

   return result;
}

bool uniqueDirectories(LogWrapperType,
                       std::string const& _temp_dir,
                       std::string const& _work_dir,
                       std::string const& _seed_dir,
                       std::string const& _dest_dir)
{
   std::vector<std::string> check_directories;
   check_directories.push_back(_temp_dir);
   check_directories.push_back(_work_dir);
   check_directories.push_back(_seed_dir);
   check_directories.push_back(_dest_dir);

   const t_uint orgSize = check_directories.size();

   std::sort(check_directories.begin(), check_directories.end());
   std::unique(check_directories.begin(), check_directories.end());

   if (check_directories.size() != orgSize)
      {
         return false;
      }

   return true;
}
